pmd-java/src/main/resources/category/java/codestyle.xml

```diff
@@ -1353,6 +1353,55 @@ public class ClassInDefaultPackage {
         </example>
     </rule>
 
+    <rule name="UseUnderscoresInNumericLiterals"
+          language="java"
+          since="6.10.0"
+          minimumLanguageVersion="1.7"
+          message="Numeric literals with more than 3 digits must use '_' as a separator"
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_codestyle.html#useunderscoresinnumericliterals">
+        <description>
+            Numeric literals with more than 3 digits must use '_' as a separator.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="version" value="2.0"/>
+            <property name="acceptableDecimalLength" type="Integer" value="4" min="3" max="6"
+                      description="Maximum acceptable length within which the rule is not applicable"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//Literal[@IntLiteral = true() or
+ @LongLiteral = true() or
+ @DoubleLiteral = true() or
+ @FloatLiteral = true()]
+ [ not (matches(@Image, "^0([xb]?[0-9a-fA-F]+)?$"))]
+ [
+  some $num in tokenize(@Image, "[.dDfFlL]|[eE][\+-]?")
+  satisfies not(
+                 string-length($num) <= $acceptableDecimalLength
+                   and (
+                         not(contains($num,"_"))
+                         or matches($num, "^[0-9]{1,3}(_[0-9]{3})*$")
```

Comment:
this line can probably be dropped...
checking "if under max length and has no underscore OR matches the regexp" should already cover all scenarios without duplicating the regexp